### PR TITLE
test: Verify if wallet is compiled in rpc_invalid_address_message.py test

### PR DESCRIPTION
**Pull request by lsilva01**
Most of  `test/functional/rpc_invalid_address_message.py` does not requires wallet.
But if the project is compiled in disable-wallet mode, the entire test will be skipped.

This PR changes the test to run the RPC tests first and then checks if the wallet is compiled.

### CODE REVIEW

File: test/functional/rpc_invalid_address_message.py [hunk 1]
```diff
@@ -2,19 +2,17 @@
...
-    assert_raises_rpc_error,
 )
 
 BECH32_VALID = 'bcrt1qtmp74ayg7p24uslctssvjm06q5phz4yrxucgnv'
 BECH32_INVALID_BECH32 = 'bcrt1p0xlxvlhemja6c4dqv22uapctqupfhlxm9h8z3k2e72q4k9hcz7vqdmchcc'
 BECH32_INVALID_BECH32M = 'bcrt1qw508d6qejxtdg4y5r3zarvary0c5xw7k35mrzd'
-BECH32_INVALID_VERSION = 'bcrt130xlxvlhemja6c4dqv22uapctqupfhlxm9h8z3k2e72q4k9hcz7vqynjegk'
 BECH32_INVALID_SIZE = 'bcrt1s0xlxvlhemja6c4dqv22uapctqupfhlxm9h8z3k2e72q4k9hcz7v8n0nx0muaewav25430mtr'
 BECH32_INVALID_V0_SIZE = 'bcrt1qw508d6qejxtdg4y5r3zarvary0c5xw7kqqq5k3my'
 BECH32_INVALID_PREFIX = 'bc1pw508d6qejxtdg4y5r3zarvary0c5xw7kw508d6qejxtdg4y5r3zarvary0c5xw7k7grplx'
```
Comment (Shubhankar-Gambhir) [inline test/functional/rpc_invalid_address_message.py:18]: Test can have more invalid address types, check different type of error messages in https://github.com/bitcoin/bitcoin/blob/master/src/key_io.cpp#:~:text=CTxDestination%20DecodeDestination(const,%7D%20//%20namespace

Comment (Shubhankar-Gambhir) [inline test/functional/rpc_invalid_address_message.py:18]: Test can have more invalid address types, check for different error messages in https://github.com/bitcoin/bitcoin/blob/master/src/key_io.cpp#:~:text=CTxDestination%20DecodeDestination(const,%7D%20//%20namespace

**Comment by Shubhankar-Gambhir** [comment]
concept ACK
there are 2 subtests and only one requires wallet, this PR enables the other one to run even with disabled wallet.
although instead of splitting can add an if condition (using self.is_wallet_compiled()) in runtests.

**Comment by aitorjs** [comment]
- Makes sense to split in two file. There are 2 subtests and only one requires wallet as @Shubhankar-Gambhir sayd.

- `self.skip_if_no_wallet()` calls to [self.is_wallet_compiled()](https://github.com/bitcoin/bitcoin/blob/master/test/functional/test_framework/test_framework.py#L817). Inside bitcoin functional tests, `self.skip_if_no_wallet()` is mainly used (77 times) and self.is_wallet_compiled() (16 times). I think that `self.is_wallet_compiled()` is used for some special cases and this one is not one of them. So, `self.skip_if_no_wallet()` is better.

- ACK ([90e382b5123ae97dc8f536e847ca790e21467446](https://github.com/bitcoin/bitcoin/pull/22794/commits/90e382b5123ae97dc8f536e847ca790e21467446))

- Tests I have done:

   - Run tests on compiled bitcoin with PR changes and wallet:
   ```sh
   $ test/functional/rpc_invalid_address_message.py
   2021-08-27T19:39:18.334000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_q38ex0pf
   2021-08-27T19:39:18.648000Z TestFramework (INFO): Stopping nodes
   2021-08-27T19:39:18.803000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_q38ex0pf on exit
   2021-08-27T19:39:18.803000Z TestFramework (INFO): Tests successful

   $ test/functional/wallet_invalid_address_message.py
   2021-08-27T19:40:12.106000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_f3rhiwtn
   2021-08-27T19:40:12.563000Z TestFramework (INFO): Stopping nodes
  2021-08-27T19:40:12.669000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_f3rhiwtn on exit
  2021-08-27T19:40:12.669000Z TestFramework (INFO): Tests successful
   ```
   - Run tests on compiled bitcoin with PR changes but without wallet (--disable-wallet):
   ```sh
   $ test/functional/rpc_invalid_address_message.py
   2021-08-27T19:34:27.769000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_ahdk12up
   2021-08-27T19:34:28.115000Z TestFramework (INFO): Stopping nodes
   2021-08-27T19:34:28.271000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_ahdk12up on exit
   2021-08-27T19:34:28.271000Z TestFramework (INFO): Tests successful

   $ test/functional/wallet_invalid_address_message.py
   2021-08-27T19:34:38.719000Z TestFramework (WARNING): Test Skipped: wallet has not been compiled.
   2021-08-27T19:34:38.769000Z TestFramework (INFO): Stopping nodes
   2021-08-27T19:34:38.769000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_2hs3yym1 on exit
   2021-08-27T19:34:38.769000Z TestFramework (INFO): Test skipped
   ```
   - Run tests on compiled bitcoin from master without wallet (--disable-wallet):
   ```sh
   $ test/functional/rpc_invalid_address_message.py
   2021-08-28T23:38:12.242000Z TestFramework (WARNING): Test Skipped: wallet has not been compiled.
   2021-08-28T23:38:12.294000Z TestFramework (INFO): Stopping nodes
   2021-08-28T23:38:12.294000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_mif2y1xy on exit
   2021-08-28T23:38:12.294000Z TestFramework (INFO): Test skipped

   $ test/functional/wallet_invalid_address_message.py
   bash: test/functional/wallet_invalid_address_message.py: File or folder does not exist
  ```

**Comment by josibake** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/22794/commits/90e382b5123ae97dc8f536e847ca790e21467446

ran the tests with and without wallet compiled to verify. you might want to indicate that this is a "move only" PR in the commit message/PR description. it helps reviewers know what they should be looking for

**Review by josibake — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/22794/commits/90e382b5123ae97dc8f536e847ca790e21467446

verified the correct tests run as expected with and without wallet compiled. you may want to add a note indicating this is "move only" to the commit or pr description. it helps reviewers know what they should be looking for

File: test/functional/rpc_invalid_address_message.py [hunk 2]
```diff
@@ -2,19 +2,17 @@
...
 
 from test_framework.util import (
     assert_equal,
-    assert_raises_rpc_error,
 )
 
 BECH32_VALID = 'bcrt1qtmp74ayg7p24uslctssvjm06q5phz4yrxucgnv'
 BECH32_INVALID_BECH32 = 'bcrt1p0xlxvlhemja6c4dqv22uapctqupfhlxm9h8z3k2e72q4k9hcz7vqdmchcc'
 BECH32_INVALID_BECH32M = 'bcrt1qw508d6qejxtdg4y5r3zarvary0c5xw7k35mrzd'
-BECH32_INVALID_VERSION = 'bcrt130xlxvlhemja6c4dqv22uapctqupfhlxm9h8z3k2e72q4k9hcz7vqynjegk'
```
Comment (MarcoFalke) [inline test/functional/rpc_invalid_address_message.py:17]: Instead of removing this, I think a test should added instead?

**Review by maflcko — COMMENTED** [review]
I think it is acceptable to use `is_wallet_compiled` here to conditionally execute the tests. For test coverage it will be no difference.

Downside is that wallet-rpcs are mixed with non-wallet-rpcs in the same file, but the benefit would be that the same node can be used to run the tests without restart (should be faster).

**Comment by Shubhankar-Gambhir** [comment]
Just in case i was not clear earlier the below patch would achieve the same 
<details>
<summary>patch</summary>

```diff
diff --git a/test/functional/rpc_invalid_address_message.py b/test/functional/rpc_invalid_address_message.py
index e362642f0..6b92b2fe1 100755
--- a/test/functional/rpc_invalid_address_message.py
+++ b/test/functional/rpc_invalid_address_message.py
@@ -29,9 +29,6 @@ class InvalidAddressErrorMessageTest(BitcoinTestFramework):
         self.setup_clean_chain = True
         self.num_nodes = 1
 
-    def skip_test_if_missing_module(self):
-        self.skip_if_no_wallet()
-
     def test_validateaddress(self):
         node = self.nodes[0]
 
@@ -87,7 +84,10 @@ class InvalidAddressErrorMessageTest(BitcoinTestFramework):
 
     def run_test(self):
         self.test_validateaddress()
-        self.test_getaddressinfo()
+
+        # skip if wallet is disabled
+        if self.is_wallet_compiled():
+            self.test_getaddressinfo()
 
 
 if __name__ == '__main__':
```
</details>

File: test/functional/rpc_invalid_address_message.py (hunk 2 above)
Comment (lsilva01) [reply test/functional/rpc_invalid_address_message.py]: Sure. Test added.

File: test/functional/rpc_invalid_address_message.py (hunk 1 above)
Comment (lsilva01) [reply test/functional/rpc_invalid_address_message.py]: Thanks. Test added.

**Comment by lsilva01** [comment]
The test has been changed to use `is_wallet_compiled` and `init_wallet`.

Also added test for `BECH32_INVALID_VERSION` case.

**Comment by Shubhankar-Gambhir** [comment]
Approach ACK
test was succesful when run on bitcoin-core compiled without wallet (--disable-wallet):
```
2021-09-20T19:30:10.881000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_bha1iotz
2021-09-20T19:30:26.999000Z TestFramework (INFO): Stopping nodes
2021-09-20T19:30:33.380000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_bha1iotz on exit
2021-09-20T19:30:33.380000Z TestFramework (INFO): Tests successful
```

**Review by Shubhankar-Gambhir — APPROVED** [review]

File: test/functional/rpc_invalid_address_message.py [hunk 3]
```diff
@@ -87,7 +84,11 @@ def test_getaddressinfo(self):
 
     def run_test(self):
         self.test_validateaddress()
-        self.test_getaddressinfo()
+
+        # skip if wallet is disabled
```
Comment (MarcoFalke) [inline test/functional/rpc_invalid_address_message.py:92]: nit: I don't think the comment is needed, when the code is self-explanatory

**Review by maflcko — COMMENTED** [review]
Needs OP/Title adjusted before merge

**Review by maflcko — APPROVED** [review]

**Comment by pg156** [comment]
Concept and approach ACK

**Comment by pg156** [comment]
Out of curiosity, how did you notice the non-wallet-rpc test is skipped when wallet is not enabled (before your PR)? Thanks @lsilva01.

**Review by jonatack — COMMENTED** [review]
Untested light review ACK f9cf67428ec25f98559a72a2504452e4b994acd7

(modulo https://github.com/bitcoin/bitcoin/pull/22794#pullrequestreview-759390259)

File: test/functional/rpc_invalid_address_message.py (hunk 3 above)
Comment (lsilva01) [reply test/functional/rpc_invalid_address_message.py]: Really, the comment is not necessary. Removed.

**Comment by lsilva01** [comment]
@p16n The code below always skips the test if the wallet is not enabled.

```
def skip_test_if_missing_module(self):
        self.skip_if_no_wallet()
```

You can verify this in the `test/functional/test_framework.py/BitcoinTestFramework/skip_if_no_wallet` file.

```
def skip_if_no_wallet(self):
        """Skip the running test if wallet has not been compiled."""
        self.requires_wallet = True
        if not self.is_wallet_compiled():
            raise SkipTest("wallet has not been compiled.")
        if self.options.descriptors:
            self.skip_if_no_sqlite()
        else:
            self.skip_if_no_bdb()
```

File: test/functional/rpc_invalid_address_message.py [hunk 4]
```diff
@@ -87,7 +84,11 @@ def test_getaddressinfo(self):
...
+        if self.is_wallet_compiled():
+            self.init_wallet(0)
```
[the remark this answers is not in the document]
Comment (lsilva01) [reply test/functional/rpc_invalid_address_message.py:93]: Interesting. I will open a new PR for this change.

**Review by maflcko — APPROVED** [review]

**Review by stratospher — COMMENTED** [review]
tested ACK c2fbdca

rpc_invalid_address_message.py checks whether correct error messages are generated for invalid addresses. Since [getaddressinfo](https://github.com/bitcoin/bitcoin/blob/5b7210c8745d9572fe94620f848d4ee1304c91a7/src/wallet/rpcwallet.cpp#L4847) RPC requires a wallet and [validateaddress](https://github.com/bitcoin/bitcoin/blob/5b7210c8745d9572fe94620f848d4ee1304c91a7/src/rpc/misc.cpp#L788) RPC does not require a wallet, the tests which invoke validateaddress should work even when it’s compiled without a wallet. This PR fixes the problem by always running the tests which invoke validateaddress and running the tests which invoke getaddressinfo only when the wallet is compiled.

Test results when compiled in disable-wallet mode:
<details>
<summary>On master</summary>
<br>

```
$ test/functional/rpc_invalid_address_message.py
2021-10-11T08:20:44.641000Z TestFramework (WARNING): Test Skipped: wallet has not been compiled.
2021-10-11T08:20:44.695000Z TestFramework (INFO): Stopping nodes
2021-10-11T08:20:44.696000Z TestFramework (INFO): Cleaning up /var/folders/bh/7x61zw991x15d01dlnx_cgrh0000gn/T/bitcoin_func_test__2uih3yl on exit
2021-10-11T08:20:44.696000Z TestFramework (INFO): Test skipped
```
</details>


<details>
<summary>On  c2fbdca</summary>
<br>

```
$ test/functional/rpc_invalid_address_message.py
2021-10-11T08:55:10.132000Z TestFramework (INFO): Initializing test directory /var/folders/bh/7x61zw991x15d01dlnx_cgrh0000gn/T/bitcoin_func_test_v7gdmep5
2021-10-11T08:55:10.994000Z TestFramework (INFO): Stopping nodes
2021-10-11T08:55:11.149000Z TestFramework (INFO): Cleaning up /var/folders/bh/7x61zw991x15d01dlnx_cgrh0000gn/T/bitcoin_func_test_v7gdmep5 on exit
2021-10-11T08:55:11.149000Z TestFramework (INFO): Tests successful
```
</details>

**Pull request merged**